### PR TITLE
fixes url encoding when no files are present

### PR DESCRIPTION
**Pull request by casualjim**

### CODE REVIEW

File: client/request.go [hunk 1]
```diff
@@ -74,6 +74,13 @@ var (
 	_ runtime.ClientRequest = new(request)
 )
 
+func (r *request) isMultipart(mediaType string) bool {
+	forFormMine := runtime.MultipartFormMime == mediaType && (len(r.formFields) > 0 || len(r.fileFields) > 0)
+	forURLMime := runtime.URLencodedFormMime == mediaType && len(r.fileFields) > 0
+	becauseFiles := runtime.MultipartFormMime != mediaType && runtime.URLencodedFormMime != mediaType && len(r.fileFields) > 0
+	return forFormMine || forURLMime || becauseFiles
+}
```
Comment (gregmarr) [inline client/request.go]: I was having trouble following this function, so I rewrote it like this:

    func (r *request) isMultipart(mediaType string) bool {
        switch mediaType {
	    case runtime.MultipartFormMime:
            return len(r.formFields) > 0 || len(r.fileFields) > 0
	    case runtime.URLencodedFormMime:
            return len(r.fileFields) > 0
	    default:
            return len(r.fileFields) > 0
    }

This lead me to no special handling for URLencodedFormMime, and one more rewrite.


    func (r *request) isMultipart(mediaType string) bool {
        // If there are fileFields, then it is multipart.
        if len(r.fileFields) > 0 {
            return true
        }
    
        // Anything other than MultipartFormMime is not multipart.
        if mediaType != runtime.MultipartFormMime {
            return false
        }
    
        // Only multipart if there are form fields.
        return len(r.formFields) > 0
    }

This seems reasonable.  I'm going to need to look to see what case this was handling.

Comment (gregmarr) [reply client/request.go]: https://github.com/go-openapi/runtime/commit/da1ede489cda53f86a098680c279a57380feee36

The fix here was making `URLencodedFormMime` equivalent to `MultipartFormMime` in terms of checking `formFields`, so I don't think this is correct.

I think you want this:

    func (r *request) isMultipart(mediaType string) bool {
        switch mediaType {
        case runtime.MultipartFormMime, runtime.URLencodedFormMime:
            return len(r.formFields) > 0 || len(r.fileFields) > 0
    	default:
            return len(r.fileFields) > 0
        }
    }

Comment (casualjim) [reply client/request.go]: URLEncodedForm should only return true when filefields are present, but not when only form fields are present.

Comment (casualjim) [reply client/request.go]: I've simplified the guard based on your feedback

Comment (gregmarr) [reply client/request.go]: If that were true, then https://github.com/go-openapi/runtime/commit/da1ede489cda53f86a098680c279a57380feee36 would not have been necessary, as that was the previous state of the code.  This commit changed it to do it for form fields too.

Comment (gregmarr) [reply client/request.go]: So the question then is, was this really required, or was the code fine before?  I'm guessing that there was some reason for this, but looking at the code, I can't see what it is.

**Comment by gregmarr** [comment]
I think this issue was with creating a client for a path that uses URL encoding and has a file property in the form body.  So maybe the check needs to look through the form fields for a file when it's URL encoding.

**Comment by casualjim** [comment]
I think there was a but in the first of the 2 if conditions which led to adding the urlencoded in, turning the bug in a new bug

File: client/request.go (hunk 1 above)
Comment (casualjim) [reply client/request.go]: this line is the reason:
https://github.com/go-openapi/runtime/commit/da1ede489cda53f86a098680c279a57380feee36#diff-ac3eb5494dc3178ea5947441c086942bR114

in this new PR both the conditions are the same now

**Comment by gregmarr** [comment]
Okay, sounds good.  I don't have a test case for this now, but it looks right.

**Pull request merged**